disk-buffering/src/main/java/io/opentelemetry/contrib/disk/buffering/internal/storage/files/ReadableFile.java

```diff
@@ -0,0 +1,153 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.contrib.disk.buffering.internal.storage.files;
+
+import io.opentelemetry.contrib.disk.buffering.internal.StorageConfiguration;
+import io.opentelemetry.contrib.disk.buffering.internal.storage.files.reader.DelimitedProtoStreamReader;
+import io.opentelemetry.contrib.disk.buffering.internal.storage.files.reader.ReadResult;
+import io.opentelemetry.contrib.disk.buffering.internal.storage.files.reader.StreamReader;
+import io.opentelemetry.contrib.disk.buffering.internal.storage.files.utils.FileTransferUtil;
+import io.opentelemetry.contrib.disk.buffering.internal.storage.responses.ReadableResult;
+import io.opentelemetry.contrib.disk.buffering.internal.storage.utils.StorageClock;
+import java.io.BufferedInputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Function;
+import javax.annotation.Nullable;
+
+public final class ReadableFile extends StorageFile {
+  private final int originalFileSize;
+  private final StreamReader reader;
+  private final FileTransferUtil fileTransferUtil;
+  private final File temporaryFile;
+  private final StorageClock clock;
+  private final long expireTimeMillis;
+  private final AtomicBoolean isClosed = new AtomicBoolean(false);
+  private int readBytes = 0;
+  @Nullable private ReadResult unconsumedResult;
+
+  public ReadableFile(
+      File file, long createdTimeMillis, StorageClock clock, StorageConfiguration configuration)
+      throws IOException {
+    this(
+        file,
+        createdTimeMillis,
+        clock,
+        configuration,
+        DelimitedProtoStreamReader.Factory.getInstance());
+  }
+
+  public ReadableFile(
+      File file,
+      long createdTimeMillis,
+      StorageClock clock,
+      StorageConfiguration configuration,
+      StreamReader.Factory readerFactory)
+      throws IOException {
+    super(file);
+    this.clock = clock;
+    expireTimeMillis = createdTimeMillis + configuration.getMaxFileAgeForReadMillis();
+    originalFileSize = (int) file.length();
+    temporaryFile = configuration.getTemporaryFileProvider().createTemporaryFile(file.getName());
+    copyFile(file, temporaryFile);
```

Comment:
I found the copy source file to temp file, read temp file and overwrite the source after each batch logic hard to follow. Comments might help.

Reply:
Instead of copying the data around it might be more efficient to alter the file format so that it would be possible to tell which chunks have already been handled. Having some kind of header in the data file could also be useful. If it is possible that files written by older version of library are read by newer version then having a header could allow skipping data in case file format has changed.

Reply:
@zeitlinger and I also discussed some more optimized approaches as well, such as having a separate file to keep track of the bytes already read and updating that number instead. The header approach I think it's a good idea as well, especially because it'd keep everything in the same file, however, I'd have to double-check how could we update the header values at the same time the file is being read to make sure we don't lose the current position if the app gets terminated without a chance to properly close the reader. Though ultimately, since there's a lot to this PR already, and, based on a benchmark added to this specific functionality, I found out that the simple approach is quite decent in terms of performance, I was thinking it would be ideal to add these optimizations on a further PR if needed.

Reply:
Exploring these ideas in the future is fine.

Reply:
Thanks, I've added an explanation in the class's doc